### DISCUSSION TITLE
WebDAV: Issues connecting with Windows

### DISCUSSION THREAD

**Discussion by tr1plus** [photoprism/photoprism#1662] [Q&A] [answered]
I am trying to connect to the "Imports" location via WebDav. I am able to do so via winscp (so the server is reachable & working), however via the windows file explorer & my QNAP NAS it is not working.

This is the URL shown in the PhotoPrism UI: http://admin@docker.triplusnet.home:2342/originals/. I have also tried without the "admin@" part added and via IP instead of hostname.  I know I can change the url in the above to be /imports/ instead of /originals/.

NOTE: Adding my nextcloud server as a "Sync" location is working fine. I'm talking about using the PhotoPrism WebDav server to add new files to the import folder.

Doing some of my own research I assume that on Windows it is probably not working due to using a non-standard http port (80/443). (but cannot really confirm that 100%).
I did confirm that on my QNAP NAS non-standard ports are allowed by adding another WebDAV Location, which was working great. As I couldn't get this to work, I first created a ticket for the NAS manufacturer, where they mentioned after some testing together with me: "Yes i agree it must be a compatibility issue that its slightly different from the standards that the NAS demands".

Because of that I want to understand if PhotoPrism is following the correct standards for WebDav or if I am missing anything. Or if there is anything else I could try. I read WebDav over NGINX is not really working so I'm not sure if it's smart to try that.

In case there would be another solution to my problem (and how I found out about this issue) - I am mainly trying to import files from a WebDav (nextcloud) location to PhotoPrism, while deleting them on nextcloud & moving them to my NAS via the Photoprism indexing. All Originals are successfully stored on my NAS already, and I'm able to import files from nextCloud via the sync, but right now they are not getting removed(as this is not working via the "sync" WebDav. this is normal, it's a "sync" not a "move"). In order to do this I wanted to move the files from WebDav Nextcloud to WebDav photoprism via a job on my NAS. This is where I am running into the above issue.

I hope this is somewhat clear. In short, the WebDav server seems to be rather difficult to connect to, depending what WebDav client you use.

Thanks!
Joren

**Comment by lastzero** [comment]
We're using the standard Go WebDAV library: https://cs.opensource.google/go/x/net/+/master:webdav/

I'd be very surprised if it doesn't comply to standards. We never experienced any issues with this.

It's true some clients can't connect via HTTP - and that is typically for security reasons as your password would be exposed:

https://docs.photoprism.org/user-guide/faq/#my-file-sync-app-fails-with-unable-to-parse-tls-packet-headers-when-trying-to-connect-via-webdav

The solution is to run Photoprism behind a HTTPS reverse proxy as advised by our docs. We recommend using Traefik.

**Comment by tr1plus** [comment]
Thanks

HTTP does work on my NAS too, so also that shouldn't be an issue. Not sure about windows tho.

Is NGINX also possible (as I already have it setup for other services)? I read that nginx WebDav support is not the best.
This should cover both the https & port issue (but both have been confirmed to not be the issue using another WebDav server).

**Comment by lastzero** [comment]
Nginx may work, but it's hard to configure properly. I know for sure chunked transfers don't work when running PHP behind nginx. but that's more of an issue for Nextcloud as we don't use PHP. The nginx reverse proxy might support it.

**Comment by tr1plus** [comment]
Okay thanks - I'll try it out later. Keep you posted.
Still a bit confused why another webdav server on the same protocol & same port is working, but not the one from Photoprism...

**Comment by lastzero** [comment]
Think we've tested Windows a few months ago and didn't have issues. Shall we test it again?

You may have to update the Basic Authentication Level in the Windows registry, as explained here at the bottom:

https://help.dreamhost.com/hc/en-us/articles/216473357-Accessing-WebDAV-with-Windows

**Comment by tr1plus** [comment]
Hi

So I've just tried again on windows with changing the registry setting. I have restarted my computer after making the change. No luck (sorry it's in Dutch)
![image](https://user-images.githubusercontent.com/7044850/130128241-cab47074-80ad-4286-a5f9-e241708123e4.png)

I will try setting up nginx tomorrow or during the week-end.

**Comment by tr1plus** [comment]
Hi - apologies for the delay.

I just configured my NGINX to also reverse proxy PhotoPrism.
Still same issue, however windows does sometimes ask for a password. (but gives the same error afterwards)
(I'm sure my SSL config for NGINX works, I use it for all my other self-hosted sites)

I also tried over http in my NGINX config. Exact same result.

Again - no issue connecting via WinSCP.

![image](https://user-images.githubusercontent.com/7044850/130662907-7c741d29-36c4-448a-899c-a51aa8a73f27.png)

```
## PhotoPrism
server {
	# SSL configuration
	listen 443 ssl;
	server_name photoprism.jump.triplusnet.home;
	
	# SSL certificates
	ssl_certificate /etc/nginx/cert/*.jump.triplusnet.home.crt;
	ssl_certificate_key /etc/nginx/cert/jump.triplusnet.home.pem;

	include /etc/nginx/includes/organizr_auth;
	
	# Reverse Proxy
	location / {
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header Host $host;

      proxy_pass http://docker.triplusnet.home:2342;

      proxy_buffering off;
      proxy_http_version 1.1;
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection "upgrade";
    }
}
```

**Comment by lastzero** [comment]
Are these self-signed certificates? If so, can you try Let's Encrypt instead?

**Comment by tr1plus** [comment]
Hi

I briefly replaced my nextcloud nginx config with the photoprism nginx config. It's public facing so I'm doing some hiding :)
Going to revert now to nextcloud, since I use that often. I can always swap back & forth now.
Nextcloud has a working webdav server on the same url/ssl certifcates.

![image](https://user-images.githubusercontent.com/7044850/131246286-1e6a1824-84df-488a-874a-760090c4f297.png)
![image](https://user-images.githubusercontent.com/7044850/131246309-66e9a8a1-ca25-4ac7-8f59-da65e7e1131c.png)
![image](https://user-images.githubusercontent.com/7044850/131246325-e8066469-4721-492a-8762-691a11174dfe.png)

**Comment by tr1plus** [comment]
I thought I had an alternative by using and FTP docker to add files to the "import" folder, but it seems the auto-import is only triggered after a change over WebDAV (as per [this](https://docs.photoprism.org/user-guide/library/import/)).

**Comment by lastzero** [comment]
Yes, we don't watch the filesystem directly as it may not support watching, or the number of watchable files may be above the limit. It also wouldn't tell us if all related files are there and complete, just provides information about single files. Photo and videos often have sidecar files. You can add a job for regular indexing / importing at your own risk (ideally at night, when you're not changing any files).

**Comment by lastzero** [comment]
Would you be able to try Traefik? Haven't used nginx or Apache for years, so I can only guess what might be the issue without spending hours replicating your setup locally. Other users can connect via WebDAV, so it's most likely not a general issue. Sorry I can't tell you more at this point.

**Comment by lastzero** [comment]
Did you check if nginx properly handles chuncked transfers? AFAIK Nextcloud doesn't support it, but it's part of the standard http://nginx.org/en/docs/http/ngx_http_core_module.html#chunked_transfer_encoding

**Comment by tr1plus** [comment]
I can try setting up traefik.
Would I be able to use this over http - without SSL(so I can keep it on my internal network). I think I should be able to get it working over port 80. Any experience with this?

**Comment by lastzero** [comment]
Some WebDAV clients won't work with HTTP: https://docs.photoprism.org/user-guide/faq/#my-file-sync-app-fails-with-unable-to-parse-tls-packet-headers-when-trying-to-connect-via-webdav

**Comment by lastzero** [comment]
Currently another nginx user with issues in our community chat, maybe you want to connect? https://gitter.im/browseyourlife/community

**Comment by lastzero** [comment]
Note that uploads and WebDAV /import/ may be disabled if you're running PhotoPrism in read-only mode as you can't add files if your originals folder is write protected. You mentioned something about imports and Nextcloud in your issue description.

**Comment by lastzero** [comment]
PhotoPrism shows an info about that if you check the logs.

**Comment by tr1plus** [comment]
> Note that uploads and WebDAV /import/ may be disabled if you're running PhotoPrism in read-only mode as you can't add files if your originals folder is write protected. You mentioned something about imports and Nextcloud in your issue description.

I was able to manually (SFTP/SSH/...) add things in the import folder, so I think that should be fine? The "Read-Only" flag is also not checked in the advanced config in PhotoPrism.



> Some WebDAV clients won't work with HTTP: https://docs.photoprism.org/user-guide/faq/#my-file-sync-app-fails-with-unable-to-parse-tls-packet-headers-when-trying-to-connect-via-webdav

I will try to play around with some other reverse proxies later. Do you know by any chance if windows "should" be able to support webdav over http? e.g. are you, or other users, able to?

**Comment by lastzero** [comment]
Windows depends on registry settings from what I know... I'd expect newer versions to have HTTP disabled by default.

**Comment by tr1plus** [comment]
okay - will keep you posted how it goes.
Thanks already for your help!

**Comment by tr1plus** [comment]
Hi

So I set up traefik & it was still not working. As a matter of test, I set up the most random docker container with webdav server(https://hub.docker.com/r/bytemark/webdav/). 
Same docker machine, same traefik reverse proxy. The webdav docker is working, photoprism is not...

![image](https://user-images.githubusercontent.com/7044850/131382575-8450e035-1a1f-41b5-8e66-2a723f5b20e5.png)

I really have a feeling something weird is happening with the photoprism webdav server

Thanks!

**Comment by tr1plus** [comment]
Hi - just want to check if there is anything more to be done here? I think I've exhausted almost every test that can be done on my side showing that it is linked to photoprism (or at least the library being used by photoprism). If you think there is anything else to test, please let me know.

**Comment by Akruidenberg** [comment]
Same problem for me. Photosync will connect, Windows 10 don't.

**Comment by Akruidenberg** [comment]
@tr1plus did some research, and WinSCP is working very well on Windows. Also, Linux (KDE Neon), and PhotoSync are working great. However, Windows 10 won't connect. I've tried multiple Windows machine with multiple versions, but no luk.
Changing the suggested registery key in Windows dont't  work. @lastzero would you take a look at this? Webdav is really important for me.

**Comment by lastzero** [comment]
I've tested it with Windows a few months ago and had no issues. Need a short break now after working that much. Can test it again then. But honestly, I think Microsoft should debug this and provide proper support. It's their product, all other clients work

**Comment by graciousgrey** [comment]
Will are going to have a look as soon as possible: https://github.com/photoprism/photoprism/issues/1665

**Comment by graciousgrey** [comment, marked as answer]
> The same steps mentioned in the documentation but with a small change top the WebDAV URL worked for me.
I used this URL at the map network drive prompt: \\photoprism.example.com@SSL\originals\

A user gave us this hint. For me it is working with \\\example.com@SSL\originals\

@tr1plus @Akruidenberg Does this work for you?

**Reply by timki0** [reply]
I've been having the same problem as @tr1plus ever since I started using PhotoPrism. @graciousgrey's suggestion didn't quite work for me until I added an extra backslash at the start. So what worked for me was \\\example.com@SSL\originals. Hope this helps others experiencing the WebDAV issues in Windows. 
Maybe something can be added to the "Connect via WebDAV" dialog that shows this different format for Windows?

EDIT: I see after I've posted that GitHub removes one of the backslashes when there are two. So I imagine @graciousgrey had the URL correct too. To be clear there needs to be two backslashes at the beginning of the URL.

**Reply by graciousgrey** [reply]
You are right, two slashes are needed. And good point regarding the dialog :+1:

**Comment by edhelll** [comment]
I had the same problem on Windows 7 with http (port 80). I have tried different solutions: change keys by regedit, change group policies, change port to 80... Finally I install KB907306 (Web Folders) from cnet.com and drive connected successfully (via drive mapping with address \\domain\originals\).
